### PR TITLE
【paddle.fleet】add gradient Merge optimizer to meta

### PR DESCRIPTION
**Pull request by mapingshuo**
### PR types

New features
### PR changes

APIs
### Describe

add gradient Merge optimizer to meta

### Test on Bert-Large model

```python
import os

os.environ['FLAGS_enable_parallel_graph'] = "0"
os.environ['FLAGS_fraction_of_gpu_memory_to_use'] = "0.98"
os.environ['FLAGS_sync_nccl_allreduce'] = "1"
os.environ['FLAGS_eager_delete_tensor_gb'] = "0"
os.environ['FLAGS_fuse_parameter_memory_size'] = "32"
os.environ['FLAGS_fuse_parameter_groups_size'] = "50"

import numpy as np
import fleet_lightning as lighting
import paddle.fluid as fluid
import paddle.fluid.incubate.fleet.base.role_maker as role_maker
import time
import paddle.fleet as fleet
import paddle

# lightning help users to focus more on learning to train a large scale model
# if you want to learn how to write a model, lightning is not for you
# focus more on engineering staff in fleet-lightning

configs = lighting.parse_train_configs()
role = role_maker.PaddleCloudRoleMaker(is_collective=True)
fleet.init(role)
# load Bert_large / Bert_base model
model = lighting.applications.Bert_large()
#model = lighting.applications.Bert_base()

data_loader = model.load_digital_dataset_from_file(
    data_dir='train/',
    vocab_path='/vocab.txt',
    max_seq_len=512,
    batch_size=14,
)

place = fluid.CUDAPlace(int(os.environ.get('FLAGS_selected_gpus', 0)))
exec_strategy = fluid.ExecutionStrategy()
exec_strategy.num_threads = 2
exec_strategy.num_iteration_per_drop_scope = 1
dist_strategy = fleet.DistributedStrategy()
dist_strategy.exec_strategy = exec_strategy
dist_strategy.nccl_comm_num = 3
dist_strategy.gradient_merge = True
dist_strategy.gradient_merge_configs = {"k_steps": 4, "avg": True}
print("configs: ", dist_strategy.gradient_merge_configs)

optimizer = fluid.optimizer.Adam(learning_rate=configs.lr)
optimizer = fleet.distributed_optimizer(optimizer, dist_strategy)
optimizer.minimize(model.loss)

print(optimizer)
print("after minimize")
exe = fluid.Executor(place)
exe.run(fluid.default_startup_program())

with open("main_program.txt", 'w') as f:
    f.write(str(paddle.default_main_program()))

scope = fluid.global_scope()
total_time = 0
for i, data in enumerate(data_loader()):
    if i >= 10:
        start_time = time.time()
    cost_val = exe.run(paddle.default_main_program(),
                       feed=data,
                       fetch_list=[model.loss.name])
    if i >= 10:
        end_time = time.time()
        total_time += (end_time - start_time)
        print(
            "worker_index: %d, step%d cost = %f, total time cost = %f, step per second: %f, speed: %f"
            % (fleet.worker_index(), i, cost_val[0], total_time,
               (i - 9) / total_time, 1 / (end_time - start_time)))
    print("step: %d, encoder_layer_16_ffn_fc_1.w_0: %s" % (
         i, scope.var("encoder_layer_16_ffn_fc_1.w_0").get_tensor().__array__()))
```

Test results:

The training results show that the GradientMerge optimizer is applied as expected.

### CODE REVIEW

File: python/paddle/fleet/base/distributed_strategy.py [hunk 1]
```diff
@@ -520,21 +520,56 @@ def dgc_configs(self, configs):
...
     @property
     def gradient_merge(self):
+        """
+        return the distribute_strategy.gradient_merge
+        """
         return self.strategy.gradient_merge
 
     @gradient_merge.setter
     def gradient_merge(self, flag):
+        """
```
Comment (guru4elephant) [inline python/paddle/fleet/base/distributed_strategy.py]: Please place these docs under property method.

File: python/paddle/fleet/base/distributed_strategy.py [hunk 2]
```diff
@@ -520,21 +520,56 @@ def dgc_configs(self, configs):
...
+        Gradient Merge, also called as Gradient Accumulation,
+        is a training strategy for larger batches. With this strategy,
```
Comment (guru4elephant) [inline python/paddle/fleet/base/distributed_strategy.py]: a training strategy for larger batches -> a strategy for large batch training.

File: python/paddle/fleet/base/distributed_strategy.py [hunk 3]
```diff
@@ -520,21 +520,56 @@ def dgc_configs(self, configs):
...
+        the parameter will not be updated until specific steps.
```
Comment (guru4elephant) [inline python/paddle/fleet/base/distributed_strategy.py]: the parameter -> model parameters.

Comment (guru4elephant) [reply python/paddle/fleet/base/distributed_strategy.py]: specific steps -> user-defined steps.

File: python/paddle/fleet/base/distributed_strategy.py [hunk 4]
```diff
@@ -520,21 +520,56 @@ def dgc_configs(self, configs):
...
+        For each step, the forward network and the backward network
+        will run to calculate the gradient of the parameters.
+        For every k step, the optimization network will run,
+        applying a specific optimization method (such as SGD, Adam)
+        to the parameters.
```
Comment (guru4elephant) [inline python/paddle/fleet/base/distributed_strategy.py]: to the parameters -> to model parameters

File: python/paddle/fleet/base/meta_optimizer_factory.py [hunk 5]
```diff
@@ -13,11 +13,14 @@
...
 
 from ..meta_optimizers import RecomputeOptimizer
+from ..meta_optimizers import GradientMergeOptimizer
 from ..meta_optimizers import GraphExecutionOptimizer
 
 __all__ = ["MetaOptimizerFactory"]
 
-meta_optimizer_names = ["RecomputeOptimizer", "GraphExecutionOptimizer"]
+meta_optimizer_names = [
+    "RecomputeOptimizer", "GradientMergeOptimizer", "GraphExecutionOptimizer"
```
Comment (guru4elephant) [inline python/paddle/fleet/base/meta_optimizer_factory.py]: 换行排列，避免冲突

**Review by guru4elephant — APPROVED** [review]
LGTM
